File: warp-client/projects/warpr-lib/src/lib/infrastructure/message-splitter.ts
import { MemoryStream } from "./memory-stream";
import { MaxSafeMessageSize, MessageContent, MessageContentType, MessageFragmentSizeBits } from "./messages";

export class MessageSplitter {
  private _messageIndex: number = 0;
  private readonly _textEncoder = new TextEncoder();

  public constructor(public MaxMessageSize = MaxSafeMessageSize) {
  }

  public *SplitMessage(message: MessageContent): Generator<Uint8Array<ArrayBuffer>> {
    let content: Uint8Array;
    let contentType : MessageContentType;

    if (typeof message === 'string') {
      content = this._textEncoder.encode(message);
      contentType = MessageContentType.Text;
    }
    else{
      content = new Uint8Array(message);
      contentType = MessageContentType.Binary;
    }
    
    let messageSize = this.MaxMessageSize - 16;
    let stream = new MemoryStream(this.MaxMessageSize);

    let fragmentSizeWithFlag = messageSize | (contentType << MessageFragmentSizeBits);
    
    let position = 0;
    let fragmentIndex = 0;
    let fragments: ArrayBuffer[] = [];
    while (position < content.length) {
      let fragmentLength = Math.min(messageSize, content.length - position);

      stream.WriteUInt32(this._messageIndex);
      stream.WriteUInt32(content.length);
      stream.WriteUInt32(fragmentSizeWithFlag);
      stream.WriteUInt32(fragmentIndex++);
      stream.WriteBytes(content.subarray(position, position + fragmentLength));

      yield stream.ToArrayBufferView();
      stream.Clear();

      position += fragmentLength;
    }

    this._messageIndex++;
  }
}

File: warp-drive/Messaging/MessageSplitter.cpp
#include "warpr_includes.h"
#include "MessageSplitter.h"
#include "MessageFragmentHeader.h"

using namespace Axodox::Storage;
using namespace std;
using namespace rtc;

namespace Warpr::Messaging
{
  MessageSplitter::MessageSplitter(size_t maxMessageSize) :
    MaxMessageSize(maxMessageSize)
  {
    _buffer.reserve(maxMessageSize);
  }

  void MessageSplitter::SplitMessage(const rtc::message_variant& message, std::function<void(const rtc::binary&)> onMessage)
  {
    //Retrieve content
    span<const uint8_t> content;
    MessageContentType contentType;
    if (holds_alternative<binary>(message))
    {
      auto& binaryMessage = get<binary>(message);
      content = { reinterpret_cast<const uint8_t*>(binaryMessage.data()), binaryMessage.size() };
      contentType = MessageContentType::Binary;
    }
    else if (holds_alternative<string>(message))
    {
      auto& textMessage = get<string>(message);
      content = { reinterpret_cast<const uint8_t*>(textMessage.data()), textMessage.size() };
      contentType = MessageContentType::Text;
    }
    else
    {
      return;
    }

    //Create header
    auto maxFragmentSize = MaxMessageSize - sizeof(MessageFragmentHeader);

    MessageFragmentHeader header;
    header.MessageIndex++;
    header.MessageSize = uint32_t(content.size());
    header.ContentType(contentType);
    header.FragmentSize(uint32_t(maxFragmentSize));

    //Write fragments
    size_t position = 0u;
    while (position < content.size())
    {
      auto fragmentLength = min(maxFragmentSize, content.size() - position);

      _buffer.resize(fragmentLength + sizeof(MessageFragmentHeader));

      memcpy(_buffer.data(), &header, sizeof(MessageFragmentHeader));
      memcpy(_buffer.data() + sizeof(MessageFragmentHeader), content.data() + position, fragmentLength);

      onMessage(_buffer);
      header.FragmentIndex++;
      position += fragmentLength;
    }
  }
}